csvwriter/csv_writer.go

```diff
@@ -0,0 +1,170 @@
+)
+
+var (
+	SaveLocally     bool
+	SaveLocallyPath string
+)
+
+func fileExists(filename string) bool {
+	if _, err := os.Stat(filename); os.IsNotExist(err) {
+		return false
+	}
+
+	return true
+}
+
+func constructHeader(orderData []byte) string {
+	var o map[string]interface{} //map[string]interface{}
+
+	err := json.Unmarshal(orderData, &o)
+	if err != nil {
+		fmt.Printf("%v\n", err)
+		return ""
+	}
+
+	var keys []string
+	for k, _ := range o {
+		keys = append(keys, k)
+	}
+
+	var outHeader string = ""
+	var idx int = 0
+	for _, k := range keys {
+		outHeader += k
+
+		if idx < len(o)-1 {
+			outHeader += ","
+		}
+		idx++
+	}
+
+	return outHeader
+}
+
+func floatToString(input_num float64) string {
+	// to convert a float number to a string
+	return strconv.FormatFloat(input_num, 'f', 6, 64)
+}
+
+func valToString(val interface{}) string {
+	switch val.(type) {
+	case int:
+		return fmt.Sprint(val.(int64))
+	case float64:
+		if val.(float64) == float64(int64(val.(float64))) {
+			return fmt.Sprint(int64(val.(float64)))
+		} else {
+			return floatToString(val.(float64))
+		}
+	case string:
+		return "\"" + fmt.Sprint(val.(string)) + "\""
+	case []interface{}:
+		outStr := "\""
+		v := val.([]interface{})
+		for ii := 0; ii < len(v); ii++ {
+
+			outStr += valToString(v[ii])
+
+			if ii < len(v)-1 {
+				outStr += ","
+			} else {
+				outStr += "\""
+			}
+		}
+		return outStr
+	}
+	return ""
+}
+
+func constructEntryFromJSON(orderData []byte) string {
+	var o map[string]interface{} //map[string]interface{}
+
+	err := json.Unmarshal(orderData, &o)
+	if err != nil {
+		fmt.Printf("%v\n", err)
+		return ""
+	}
+
+	var keys []string
+	for k, _ := range o {
+		keys = append(keys, fmt.Sprint(k))
+	}
+
+	var outStr string = ""
+
+	var idx int = 0
+	for _, k := range keys {
+		fn := o[k]
+		outStr += valToString(fn)
+
+		if idx < len(o)-1 {
+			outStr += ","
+		}
+		idx++
+	}
+
+	return outStr
+}
+
+func writeDataToFile(msg *nats.Msg, identification string, filename string) bool {
+	os.MkdirAll(strings.TrimSuffix(strings.TrimSuffix(SaveLocallyPath, "/"), "\\"), os.ModePerm)
+	fExists := fileExists(filename)
+
+	file, err := os.OpenFile(filename, os.O_RDWR|os.O_APPEND|os.O_CREATE, 0660)
+	if err != nil {
+		fmt.Printf("%v\n", err)
+		return false
+	}
+
+	var entry string = ""
+	if !fExists {
+		entry = constructHeader(msg.Data)
+		if entry == "" {
+			fmt.Printf("Failed to construct header for %v\n", identification)
+			return false
+		}
+
+		file.WriteString(entry + "\n")
+	}
+
+	entry = constructEntryFromJSON(msg.Data)
+	if entry == "" {
+		fmt.Printf("Failed to create entry for %v\n", identification)
+		return false
+	}
+
+	file.WriteString(entry + "\n")
+	file.Close()
+
+	return true
+}
+
+func SaveToCSVFile(msg *nats.Msg, identification string) {
+	var filename string = strings.TrimSuffix(strings.TrimSuffix(SaveLocallyPath, "/"), "\\") + "/" + identification + "_" + time.Now().Format("20060102150405") + ".csv"
```

Comment:
See this: https://golang.org/pkg/path/filepath/#Join

It automatically joins file paths removing and adding slashes where needed.